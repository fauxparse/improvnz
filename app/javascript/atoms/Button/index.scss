@import "../../stylesheets/all.scss";

.button {
  @include button-color(blue-grey);
  @include button-size;

  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: 0;
  border-radius: 0.25rem;
  font-weight: 500;
  cursor: pointer;
  user-select: none;

  &--small {
    @include button-size(
      $font-size: 0.875rem,
      $padding: 0.25rem 0.5rem,
    );
  }

  &--large {
    @include button-size(
      $font-size: 1.25rem,
      $padding: 1rem 2rem,
    );
  }

  &--outline {
    border: 0.0625rem solid transparent;
    background-clip: border-box;
  }

  &--default {
    @include button-color(cyan);
  }

  &--special {
    @include button-color(pink);
  }

  &--success {
    @include button-color(lime);
  }

  &--warning {
    @include button-color(yellow);
  }

  &--danger {
    @include button-color(rose);
  }

  &--toolbar {
    @include button-color(blue-grey);
  }

  &--icon-left {
    flex-direction: row;
  }

  &--icon-right {
    flex-direction: row-reverse;
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  .icon {
    color: inherit;
  }
}
