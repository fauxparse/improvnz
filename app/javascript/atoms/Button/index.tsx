import React, { ButtonHTMLAttributes } from 'react';
import classNames from 'clsx';
import omit from 'lodash/omit';
import flatMap from 'lodash/flatMap';
import Icon, { NamedIcon } from '../Icon';
import './index.scss';

type OptionalProps<T extends string> = { [key in T]?: boolean };

export enum ButtonSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export enum ButtonFlavour {
  default = 'default',
  special = 'special',
  success = 'success',
  warning = 'warning',
  danger = 'danger',
  toolbar = 'toolbar',
}

export enum ButtonStyle {
  primary = 'primary',
  secondary = 'secondary',
  outline = 'outline',
  clear = 'clear',
}

type ButtonIconPosition = 'left' | 'right';

interface ButtonVariants
  extends OptionalProps<ButtonSize>,
    OptionalProps<ButtonFlavour>,
    OptionalProps<ButtonStyle> {}
export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement>, ButtonVariants {
  className?: string;
  text?: string;
  icon?: NamedIcon;
  iconPosition?: ButtonIconPosition;
}

const extract = <T extends ButtonVariants>(kind, props: T, defaultValue) =>
  Object.keys(kind).find((prop) => props[kind[prop]]) || defaultValue;

const otherPropNames: [keyof ButtonVariants][] = flatMap(
  [ButtonSize, ButtonFlavour, ButtonSize],
  Object.values
);

const otherProps = <T extends ButtonVariants>(props: T) => omit(props, otherPropNames);

const Button: React.FC<ButtonProps> = ({
  className,
  type = 'button',
  icon,
  iconPosition = 'left',
  text,
  children,
  ...props
}) => (
  <button
    className={classNames(
      'button',
      `button--${extract(ButtonSize, props, ButtonSize.medium)}`,
      `button--${extract(ButtonFlavour, props, ButtonFlavour.default)}`,
      `button--${extract(ButtonStyle, props, ButtonStyle.secondary)}`,
      icon && `button--icon-${iconPosition}`,
      className
    )}
    type={type}
    {...otherProps(props)}
  >
    {icon && <Icon className="button__icon" icon={icon} />}
    {text && <span className="button__text">{text}</span>}
    {children}
  </button>
);

Button.displayName = 'Button';

export default Button;
