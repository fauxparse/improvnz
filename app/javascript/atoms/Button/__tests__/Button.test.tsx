import React from 'react';
import { render } from '@testing-library/react';
import Button from '../';

describe('Button', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Button text="Hello" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('contains the correct text', () => {
    const { container } = render(<Button text="Hello" />);
    expect(container).toHaveTextContent('Hello');
  });

  it('has text', () => {
    const { container } = render(<Button text="Hello" />);
    expect(container.querySelector('.button__text')).toHaveTextContent('Hello');
  });

  it('has no icon', () => {
    const { container } = render(<Button text="Hello" />);
    expect(container.querySelector('.button__icon')).toBe(null);
  });

  describe('with an icon', () => {
    it('renders correctly', () => {
      const { asFragment } = render(<Button icon="home" />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('renders the icon', () => {
      const { container } = render(<Button icon="home" />);
      expect(container.querySelector('.button__icon').getAttribute('data-icon')).toBe('home');
    });

    it('does not render button text', () => {
      const { container } = render(<Button icon="home" />);
      expect(container.querySelector('.button__text')).toBe(null);
    });
  });

  describe('with additional children', () => {
    it('renders correctly', () => {
      const { asFragment } = render(<Button>Hello</Button>);
      expect(asFragment()).toMatchSnapshot();
    });

    it('renders extra children', () => {
      const { container } = render(<Button>Hello</Button>);
      expect(container).toHaveTextContent('Hello');
    });
  });
});
