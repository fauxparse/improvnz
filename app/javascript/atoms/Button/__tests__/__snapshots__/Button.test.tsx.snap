// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<DocumentFragment>
  <button
    class="button button--medium button--default button--secondary"
    type="button"
  >
    <span
      class="button__text"
    >
      Hello
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button with additional children renders correctly 1`] = `
<DocumentFragment>
  <button
    class="button button--medium button--default button--secondary"
    type="button"
  >
    Hello
  </button>
</DocumentFragment>
`;

exports[`Button with an icon renders correctly 1`] = `
<DocumentFragment>
  <button
    class="button button--medium button--default button--secondary button--icon-left"
    type="button"
  >
    <svg
      class="icon icon--outline button__icon"
      data-icon="home"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6"
      />
    </svg>
  </button>
</DocumentFragment>
`;
